common/elasticsearch/client/os2/client_bulk.go

```diff
@@ -0,0 +1,205 @@
+	case bulk.BulkableIndexRequest:
+		body, err := json.Marshal(request.Doc)
+		if err != nil {
+			v.logger.Error("marshal bulk index request doc", tag.Error(err))
+			return
+		}
+		req.Action = "index"
+		req.Body = bytes.NewReader(body)
+		callBackRequest = bulk.NewBulkIndexRequest().
+			ID(request.ID).
+			Index(request.Index).
+			Version(request.Version).
+			VersionType(request.VersionType).Doc(request.Doc)
+	case bulk.BulkableCreateRequest:
+		body, err := json.Marshal(request.Doc)
+		if err != nil {
+			v.logger.Error("marshal bulk create request doc", tag.Error(err))
+			return
+		}
+		versionType := "internal"
+		req.Action = "create"
+		req.VersionType = &versionType
+		req.Body = bytes.NewReader(body)
+		callBackRequest = bulk.NewBulkIndexRequest().ID(request.ID).
+			Index(request.Index).
+			Version(request.Version).
+			VersionType(versionType).Doc(request.Doc)
+
+	}
+
+	req.OnFailure = func(ctx context.Context, item opensearchutil.BulkIndexerItem, res opensearchutil.BulkIndexerResponseItem, err error) {
+		v.before(0, metricsRequest)
+		it := bulk.GenericBulkResponseItem{
+			Index:       res.Index,
+			ID:          res.DocumentID,
+			Version:     res.Version,
+			SeqNo:       res.SeqNo,
+			PrimaryTerm: res.PrimTerm,
+			Status:      res.Status,
+			Error:       res.Error,
+		}
+
+		gr := []bulk.GenericBulkableRequest{callBackRequest}
+		gbr := bulk.GenericBulkResponse{
+			Took:   0,
+			Errors: res.Error.Type != "" || res.Status > 201,
+			Items: []map[string]*bulk.GenericBulkResponseItem{
+				{req.Action: &it},
+			},
+		}
+
+		ge := bulk.GenericError{
+			Status:  res.Status,
+			Details: err,
+		}
+
+		v.after(0, gr, &gbr, &ge)
+
+	}
+
+	req.OnSuccess = func(ctx context.Context, item opensearchutil.BulkIndexerItem, res opensearchutil.BulkIndexerResponseItem) {
+		v.before(0, metricsRequest)
+
+		gr := []bulk.GenericBulkableRequest{callBackRequest}
+
+		it := bulk.GenericBulkResponseItem{
+			Index:       res.Index,
+			ID:          res.DocumentID,
+			Version:     res.Version,
+			SeqNo:       res.SeqNo,
+			PrimaryTerm: res.PrimTerm,
+			Status:      res.Status,
+			Error:       res.Error,
+		}
+
+		gbr := bulk.GenericBulkResponse{
+			Took:   0,
+			Errors: false,
+			Items: []map[string]*bulk.GenericBulkResponseItem{
+				{req.Action: &it},
+			},
+		}
+
+		v.after(0, gr, &gbr, nil /*No errors here*/)
+	}
+	if err := v.processor.Add(context.Background(), req); err != nil {
+		v.logger.Error("adding bulk request to OpenSearch", tag.Error(err))
+	}
+}
+
+func (v *bulkProcessor) Flush() error {
+	// Opensearch client doesn't provide flush method
+	return nil
+}
+
+func (c *OS2) RunBulkProcessor(_ context.Context, parameters *bulk.BulkProcessorParameters) (bulk.GenericBulkProcessor, error) {
+	processor, err := opensearchutil.NewBulkIndexer(opensearchutil.BulkIndexerConfig{
+		Client:        c.client,
+		FlushInterval: parameters.FlushInterval,
+		NumWorkers:    parameters.NumOfWorkers,
+		Decoder:       &NumberDecoder{},
```

Comment:
tbh I wonder if this would be better as a `(*NumberDecoder)(nil)`.  zero-sized object pointers are quite funky and sometimes error-prone: http://devs.cloudimmunity.com/gotchas-and-common-mistakes-in-go-golang/#zero_size_var